### PR TITLE
Implement ML Features #381. SQLTransformer class and testcase

### PR DESCRIPTION
**Pull request by ramanathanv**
Hi,
I am creating this draft pull request that implements the basic methods in SQLTransformer class along with Test cases. 
This is related to "Implement ML Features" #381

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/spark?pullRequest=781) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/spark?pullRequest=781) before we can accept your contribution.<br/><br/>:x: ramanathanv [sign now](https://cla.dotnetfoundation.org/dotnet/spark?pullRequest=781)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/spark?pullRequest=781) it.</sub>

File: src/csharp/Microsoft.Spark.E2ETest/IpcTests/ML/Feature/SQLTransformerTests.cs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+                string savePath = Path.Join(tempDirectory.Path, "SQLTransformer");
+                sqlTransformer.Save(savePath);
+
+                SQLTransformer loadedsqlTransformer = SQLTransformer.Load(savePath);
+                Assert.Equal(sqlTransformer.Uid(), loadedsqlTransformer.Uid());
+            }
+
+            Assert.Equal(expectedUid, sqlTransformer.Uid());
+        }
+
```
Comment (GoEddie) [inline src/csharp/Microsoft.Spark.E2ETest/IpcTests/ML/Feature/SQLTransformerTests.cs:62]: extra blank line

File: src/csharp/Microsoft.Spark/ML/Feature/SQLTransformer.cs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+        /// <summary>
+        /// Executes the <see cref="SQLTransformer"/> and transforms the DataFrame to include the new
+        /// column.
+        /// </summary>
+        /// <param name="source">The DataFrame to transform</param>
+        /// <returns>
+        /// New <see cref="DataFrame"/> object with the source <see cref="DataFrame"/> transformed
+        /// </returns>
+        public DataFrame Transform(DataFrame source) =>
```
Comment (GoEddie) [inline src/csharp/Microsoft.Spark/ML/Feature/SQLTransformer.cs:49]: For consistency, is it possible to add TransformSchema as well as Transform? https://spark.apache.org/docs/2.1.0/api/java/org/apache/spark/ml/feature/SQLTransformer.html#transformSchema(org.apache.spark.sql.types.StructType)

File: src/csharp/Microsoft.Spark/ML/Feature/SQLTransformer.cs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+            new DataFrame((JvmObjectReference)_jvmObject.Invoke("transform", source));
+
+        /// <summary>
+        /// Sets the statement to <see cref="SQLTransformer"/>.
+        /// </summary>
+        /// <param name="statement">SQL Statement</param>
+        /// <returns>
+        /// <see cref="SQLTransformer"/> with the statement set
+        /// </returns>
+        public SQLTransformer SetStatement(string statement) =>
```
Comment (GoEddie) [inline src/csharp/Microsoft.Spark/ML/Feature/SQLTransformer.cs:59]: For consistency, is it possible to add GetStatement as well as SetStatement? https://spark.apache.org/docs/2.1.0/api/java/org/apache/spark/ml/feature/SQLTransformer.html#getStatement()

Comment (ramanathanv) [reply src/csharp/Microsoft.Spark/ML/Feature/SQLTransformer.cs:79]: Hi @GoEddie . I have added TransformSchema() and GetStatement() in SQLTransformer class. Can you please check?

**Comment by ramanathanv** [comment]
Hi @GoEddie . I have implemented your review comments and have added TransformSchema() and GetStatement() in SQLTransformer class.

**Comment by ramanathanv** [comment]
Closed by mistake!

File: src/csharp/Microsoft.Spark/ML/Feature/SQLTransformer.cs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+            new StructType(
+                (JvmObjectReference)_jvmObject.Invoke(
+                    "transformSchema",
+                    DataType.FromJson(_jvmObject.Jvm, value.Json)));
+
+        /// <summary>
+        /// Gets the statement.
+        /// </summary>
+        /// <returns>Statement</returns>
+        public string GetStatement() => (string)(_jvmObject.Invoke("getStatement"));
```
Comment (GoEddie) [inline src/csharp/Microsoft.Spark/ML/Feature/SQLTransformer.cs:70]: Can you add a test for GetStatement and the parens around _jvm are not necessary (_jvmObject.Invoke("getStatement"))

File: src/csharp/Microsoft.Spark/ML/Feature/SQLTransformer.cs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        /// </returns>
+        public SQLTransformer SetStatement(string statement) =>
+            WrapAsSQLTransformer ((JvmObjectReference)_jvmObject.Invoke("setStatement", statement));
+
+        /// <summary>
+        /// Loads the <see cref="SQLTransformer"/> that was previously saved using Save.
+        /// </summary>
+        /// <param name="path">The path the previous <see cref="SQLTransformer"/> was saved to</param>
+        /// <returns>New <see cref="SQLTransformer"/> object, loaded from path</returns>
+        public static SQLTransformer Load(string path)
```
Comment (GoEddie) [inline src/csharp/Microsoft.Spark/ML/Feature/SQLTransformer.cs:87]: Maybe change this to a method group so it is consistent "(Load(string path) => return Wrap..."

**Review by GoEddie — COMMENTED** [review]
Looks good, a couple of small things

File: src/csharp/Microsoft.Spark/ML/Feature/SQLTransformer.cs (hunk 5 above)
Comment (GoEddie) [reply src/csharp/Microsoft.Spark/ML/Feature/SQLTransformer.cs:87]: Could @Niharikadutta review?

File: src/csharp/Microsoft.Spark.E2ETest/IpcTests/ML/Feature/SQLTransformerTests.cs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+                    new StructField("v2", new DoubleType())
+                }));
+
+            string expectedUid = "theUid";
+
+            SQLTransformer sqlTransformer = new SQLTransformer(expectedUid)
+                .SetStatement("SELECT *, (v1 + v2) AS v3, (v1 * v2) AS v4 FROM __THIS__");
+
+            DataFrame output = sqlTransformer.Transform(input);
+            var outputSchema = sqlTransformer.TransformSchema(input.Schema());
```
Comment (Niharikadutta) [inline src/csharp/Microsoft.Spark.E2ETest/IpcTests/ML/Feature/SQLTransformerTests.cs:47]: Let's use an explicit `StructType` type here

File: src/csharp/Microsoft.Spark.E2ETest/IpcTests/ML/Feature/SQLTransformerTests.cs [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+                string savePath = Path.Join(tempDirectory.Path, "SQLTransformer");
+                sqlTransformer.Save(savePath);
+
+                SQLTransformer loadedsqlTransformer = SQLTransformer.Load(savePath);
+                Assert.Equal(sqlTransformer.Uid(), loadedsqlTransformer.Uid());
+            }
+            Assert.Equal(expectedUid, sqlTransformer.Uid());
+        }
+
+    }
```
Comment (Niharikadutta) [inline src/csharp/Microsoft.Spark.E2ETest/IpcTests/ML/Feature/SQLTransformerTests.cs:65]:
```suggestion
        }
    }
```

File: src/csharp/Microsoft.Spark/ML/Feature/SQLTransformer.cs [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+namespace Microsoft.Spark.ML.Feature
+{
+    /// <summary>
+    /// A <see cref="SQLTransformer"/> implements the transformations which are defined by SQL statement.
+    /// </summary>
+    public class SQLTransformer : FeatureBase<SQLTransformer>, IJvmObjectReferenceProvider
+    {
+        private static readonly string s_sqlTransformerClassName =
+    "org.apache.spark.ml.feature.SQLTransformer";
```
Comment (Niharikadutta) [inline src/csharp/Microsoft.Spark/ML/Feature/SQLTransformer.cs:18]:
```suggestion
        private static readonly string s_sqlTransformerClassName = 
            "org.apache.spark.ml.feature.SQLTransformer";
```

File: src/csharp/Microsoft.Spark/ML/Feature/SQLTransformer.cs [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+        /// Loads the <see cref="SQLTransformer"/> that was previously saved using Save.
+        /// </summary>
+        /// <param name="path">The path the previous <see cref="SQLTransformer"/> was saved to</param>
+        /// <returns>New <see cref="SQLTransformer"/> object, loaded from path</returns>
+        public static SQLTransformer Load(string path)
+        {
+            return WrapAsSQLTransformer(
+                SparkEnvironment.JvmBridge.CallStaticJavaMethod(
+                    s_sqlTransformerClassName, "load", path));
+        }
```
Comment (Niharikadutta) [inline src/csharp/Microsoft.Spark/ML/Feature/SQLTransformer.cs:92]:
```suggestion
        public static SQLTransformer Load(string path) =>
            WrapAsSQLTransformer(
                SparkEnvironment.JvmBridge.CallStaticJavaMethod(
                    s_sqlTransformerClassName,
                    "load",
                    path));
```

**Review by Niharikadutta — CHANGES_REQUESTED** [review]
Added a few comments @ramanathanv . Thanks!

**Comment by ramanathanv** [comment]
@GoEddie , @Niharikadutta . Thank you for the review comments! I have implemented them in the commit now.

File: src/csharp/Microsoft.Spark/ML/Feature/SQLTransformer.cs [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+
+using Microsoft.Spark.Interop;
+using Microsoft.Spark.Interop.Ipc;
+using Microsoft.Spark.Sql;
+using Microsoft.Spark.Sql.Types;
+
+namespace Microsoft.Spark.ML.Feature
+{
+    /// <summary>
+    /// A <see cref="SQLTransformer"/> implements the transformations which are defined by SQL statement.
```
Comment (Niharikadutta) [inline src/csharp/Microsoft.Spark/ML/Feature/SQLTransformer.cs:13]:
```suggestion
    /// <see cref="SQLTransformer"/> implements the transformations which are defined by SQL statement.
```

File: src/csharp/Microsoft.Spark/ML/Feature/SQLTransformer.cs [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+
+        JvmObjectReference IJvmObjectReferenceProvider.Reference => _jvmObject;
+
+        /// <summary>
+        /// Executes the <see cref="SQLTransformer"/> and transforms the DataFrame to include the new
+        /// column.
+        /// </summary>
+        /// <param name="source">The DataFrame to transform</param>
+        /// <returns>
+        /// New <see cref="DataFrame"/> object with the source <see cref="DataFrame"/> transformed
```
Comment (Niharikadutta) [inline src/csharp/Microsoft.Spark/ML/Feature/SQLTransformer.cs:48]:
```suggestion
        /// New <see cref="DataFrame"/> object with the source <see cref="DataFrame"/> transformed.
```

File: src/csharp/Microsoft.Spark/ML/Feature/SQLTransformer.cs [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+        /// </returns>
+        public DataFrame Transform(DataFrame source) =>
+            new DataFrame((JvmObjectReference)_jvmObject.Invoke("transform", source));
+
+        /// <summary>
+        /// Executes the <see cref="SQLTransformer"/> and transforms the schema.
+        /// </summary>
+        /// <param name="value">The Schema to be transformed</param>
+        /// <returns>
+        /// New <see cref="StructType"/> object with the schema <see cref="StructType"/> transformed
```
Comment (Niharikadutta) [inline src/csharp/Microsoft.Spark/ML/Feature/SQLTransformer.cs:58]:
```suggestion
        /// New <see cref="StructType"/> object with the schema <see cref="StructType"/> transformed.
```

File: src/csharp/Microsoft.Spark/ML/Feature/SQLTransformer.cs [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+        /// </summary>
+        /// <returns>Statement</returns>
+        public string GetStatement() => (string)_jvmObject.Invoke("getStatement");
+
+        /// <summary>
+        /// Sets the statement to <see cref="SQLTransformer"/>.
+        /// </summary>
+        /// <param name="statement">SQL Statement</param>
+        /// <returns>
+        /// <see cref="SQLTransformer"/> with the statement set
```
Comment (Niharikadutta) [inline src/csharp/Microsoft.Spark/ML/Feature/SQLTransformer.cs:77]:
```suggestion
        /// <see cref="SQLTransformer"/> with the statement set.
```

File: src/csharp/Microsoft.Spark/ML/Feature/SQLTransformer.cs [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+        /// </returns>
+        public SQLTransformer SetStatement(string statement) =>
+            WrapAsSQLTransformer ((JvmObjectReference)_jvmObject.Invoke("setStatement", statement));
```
Comment (Niharikadutta) [inline src/csharp/Microsoft.Spark/ML/Feature/SQLTransformer.cs:80]:
```suggestion
            WrapAsSQLTransformer((JvmObjectReference)_jvmObject.Invoke("setStatement", statement));
```

File: src/csharp/Microsoft.Spark.E2ETest/IpcTests/ML/Feature/SQLTransformerTests.cs [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+    public class SQLTransformerTests : FeatureBaseTests<SQLTransformer>
+    {
+        private readonly SparkSession _spark;
+
+        public SQLTransformerTests(SparkFixture fixture) : base(fixture)
+        {
+            _spark = fixture.Spark;
+        }
+
+        [Fact]
```
Comment (Niharikadutta) [inline src/csharp/Microsoft.Spark.E2ETest/IpcTests/ML/Feature/SQLTransformerTests.cs:25]: nit: Can we add a small summary for this test?

**Review by Niharikadutta — COMMENTED** [review]
A few nits, but otherwise LGTM. Thanks @ramanathanv !

**Comment by ramanathanv** [comment]
Hi @GoEddie , @Niharikadutta . I have implemented the code review fixes. But I notice that the build fails and I guess it is not due to the code change. Can you please help me verify?

**Comment by Niharikadutta** [comment]
> Hi @GoEddie , @Niharikadutta . I have implemented the code review fixes. But I notice that the build fails and I guess it is not due to the code change. Can you please help me verify?

@ramanathanv Yes this is a known issue that occurs intermittently and is being tracked in #773 . I have re-triggered this pipeline to re-run all the tests. Thanks!

**Comment by ramanathanv** [comment]
> @ramanathanv Yes this is a known issue that occurs intermittently and is being tracked in #773 . I have re-triggered this pipeline to re-run all the tests. Thanks!

Thanks @Niharikadutta  and @GoEddie . The build has passed all checks now. Please let me know if there is something pending from my end for this pull request to be accepted.

**Comment by GoEddie** [comment]
There is one comment left to complete (adding a short summary of the test) and then it looks good to me:

https://github.com/dotnet/spark/pull/781/files/4fcda8904d374c52fbb97dbf7618145d0a15ca4e#r535114596

**Comment by ramanathanv** [comment]
> There is one comment left to complete (adding a short summary of the test) and then it looks good to me:
> 
> https://github.com/dotnet/spark/pull/781/files/4fcda8904d374c52fbb97dbf7618145d0a15ca4e#r535114596

Hi @GoEddie . I had included the comment earlier. But missed to "resolve" the review comment. I have now resolved it.

**Comment by GoEddie** [comment]
Hi @ramanathanv i'm not sure what happened to your changes but most of the comments have been resolved but without the code changes - maybe you have a local commit that you haven't pushed up to github? (or you made them on another branch or something?)

For example:

https://github.com/dotnet/spark/pull/781/files/4fcda8904d374c52fbb97dbf7618145d0a15ca4e#r535114596

The comment is: "nit: Can we add a small summary for this test?" but there is no summary, the ///<summary> method headers 

or:

https://github.com/dotnet/spark/pull/781/files/4fcda8904d374c52fbb97dbf7618145d0a15ca4e#r530571957

The code style says to use `var` only when calling a constructor, in this case where you call the constructor and then a method (TransformSchema) then we should use `StructType` instead of `var`

If you look through the other comments you will see a few things to implement - if you aren't sure what to do or need a hand with them let me know and I can help!

**Comment by ramanathanv** [comment]
Hi @GoEddie .
When I navigate to the "Files Changed" submenu within this PR, I notice that the review comments are indeed fixed. I have attached a screenshot along with this conversation. Can you please recheck and let me know if you are able to see the code with the review comments fixed?

image

**Comment by GoEddie** [comment]
@ramanathanv looks good!

**Review by GoEddie — APPROVED** [review]

**Comment by GoEddie** [comment]
@Niharikadutta could you review?

**Review by imback82 — APPROVED** [review]
LGTM, thanks @ramanathanv!

**Pull request merged**